### ISSUE TITLE
Initial runtime errors with SMS_D_Ln2_P1x1.ne4_ne4.F2000SCREAMv1.perlmutter_gnugpu

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/scream#1443]
New note: the above test will currently default to `ATM_NCPL=12` which is known to be incorrect.  The right setting is either `ATM_NCPL=288` or `ATM_NCPL=48`. The test modifier `scream-30min_ts`, it looks like it's `ATM_NCPL=48`.
So the correct (or at least working) create_test name test would then be:
`SMS_D_Ld1_P1x1.ne4_ne4.F2000SCREAMv1.perlmutter_gnugpu.scream-30min_ts`
(note also changing length of time to be 1 day as my errors with default NCPL=12 were after step 2)

On Perlmutter, with 1 MPI I get about 69k messages of the form below. I don't know of they are related to the assertion fail or not.

I was going to make a separate issue for P2x1, but while it fails in a different way, it does seem to stop at the same point (same place with cpl.log and both have zero atm and lnd log files).

```
0:       M =   0.100469287022012E+05 kg/m^2  0.985217903623769E+05mb
0: WARNING: SPA Remap File has been set to 'NONE', assuming that SPA data and simulation are on the same grid - skipping horizontal interpolationCreate Pool
0: A100-SXM4-40GB
0: KERNEL CHECK FAILED:
0:    !ekat::OnGpu<typename device_type::execution_space>::value || is_pow_of_2(team.team_size())
0:    Error! Team-level parallel_scan on CUDA only works for team sizes that are power of 2.
0:        You could try to reduce the team size to the previous pow of 2.
0:
0: KERNEL CHECK FAILED:
0:    !ekat::OnGpu<typename device_type::execution_space>::value || is_pow_of_2(team.team_size())
0:    Error! Team-level parallel_scan on CUDA only works for team sizes that are power of 2.
0:        You could try to reduce the team size to the previous pow of 2.

...

0: :0: : block: [203,0,0], thread: [0,0,0] Assertion `` failed.
0: :0: : block: [203,0,0], thread: [0,1,0] Assertion `` failed.
0: :0: : block: [203,0,0], thread: [0,2,0] Assertion `` failed.

...

0: terminate called after throwing an instance of 'std::runtime_error'
0:   what():  cudaStreamSynchronize(m_stream) error( cudaErrorAssert): device-side assert triggered /pscratch/sd/n/ndk/wacmy/s12-mar1/externals/ekat/extern/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:314
0: Traceback functionality not available
0:
0:
0: Program received signal SIGABRT: Process abort signal.
0:
0: Backtrace for this error:
0: #0  0x1504b6fb149f in ???
0: #1  0x1504b6fb1420 in ???
0: #2  0x1504b6fb2a00 in ???
0: #3  0x1504bd31af59 in _ZN9__gnu_cxx27__verbose_terminate_handlerEv
0:      at ../../../../cpe-gcc-11.2.0-202108140355.9bf1fd589a5c1/libstdc++-v3/libsupc++/vterminate.cc:95
0: #4  0x1504bd3267b9 in _ZN10__cxxabiv111__terminateEPFvvE
0:      at ../../../../cpe-gcc-11.2.0-202108140355.9bf1fd589a5c1/libstdc++-v3/libsupc++/eh_terminate.cc:48
0: #5  0x1504bd326824 in _ZSt9terminatev
0:      at ../../../../cpe-gcc-11.2.0-202108140355.9bf1fd589a5c1/libstdc++-v3/libsupc++/eh_terminate.cc:58
0: #6  0x1504bd326ab7 in __cxa_throw
0:      at ../../../../cpe-gcc-11.2.0-202108140355.9bf1fd589a5c1/libstdc++-v3/libsupc++/eh_throw.cc:95
0: #7  0x31ffa68 in _ZN6Kokkos4Impl23throw_runtime_exceptionERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/externals/ekat/extern/kokkos/core/src/impl/Kokkos_Error.cpp:72
0: #8  0x3209e51 in _ZN6Kokkos4Impl25cuda_internal_error_throwE9cudaErrorPKcS3_i
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/externals/ekat/extern/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:153
0: #9  0x5c7925 in _ZN6Kokkos4Impl23cuda_internal_safe_callE9cudaErrorPKcS3_i
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/externals/ekat/extern/kokkos/core/src/Cuda/Kokkos_Cuda_Error.hpp:67
0: #10  0x320a4c2 in _ZNK6Kokkos4Impl12CudaInternal5fenceEv
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/externals/ekat/extern/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:314
0: #11  0x269aa43 in ???
0: #12  0x2699897 in ???
0: #13  0x26981af in ???
0: #14  0x26962b0 in ???
0: #15  0x2692bf0 in _ZN6scream7control15SurfaceCoupling9do_exportEb
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/control/surface_coupling_export.cpp:47
0: #16  0x5c0a9b in operator()
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp:160
0: #17  0x5c5c98 in fpe_guard_wrapper<scream_setup_surface_coupling(char const*&, int const*&, double*&, int const*&, int const&, int const&, char const*&, int const*&, double*&, int const*&, bool const*&, int const&)::<lambda()> >
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp:50
0: #18  0x5c0ba1 in scream_setup_surface_coupling
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp:132
0: #19  0x5bd959 in __atm_comp_mct_MOD_atm_init_mct
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/atm_comp_mct.F90:159
0: #20  0x44337e in __component_mod_MOD_component_init_cc
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/component_mod.F90:248
0: #21  0x42b1d2 in __cime_comp_mod_MOD_cime_init
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_comp_mod.F90:1425
0: #22  0x43c5ab in cime_driver
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_driver.F90:122
0: #23  0x43c6eb in main
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_driver.F90:23

casedir:
/pscratch/sd/n/ndk/wacmy/s12-mar1/SMS_D_Ln2_P1x1.ne4_ne4.F2000SCREAMv1.perlmutter_gnugpu.20220301_123431_2cfypu
```

To reproduce, unfortunately, need a few updates that are not yet in master.

1) because of https://github.com/E3SM-Project/E3SM/issues/4809, need to do "module load craype-x86-rome". This will be used for all jobs in that terminal. I don't want to add this to machine file as we will be bringing in a different change with upstream merge (once e3sm master is updated)

2) The file `externals/ekat/cmake/machine-files/kokkos/nvidia-a100.cmake` is not yet in scream master as it's working it's way thru ekat.  Can simply copy this small file from
`/pscratch/sd/n/ndk/wacmy/s12-mar1/externals/ekat/cmake/machine-files/kokkos/nvidia-a100.cmake`

3) known issue https://github.com/E3SM-Project/scream/issues/1436 with at least one of the input netcdf files not being in the correct format. For now, edit directly `run/data/scream_input.yaml` and use:
`SPA Data File: /global/cfs/cdirs/e3sm/inputdata/atm/scream/init/spa_file_unified_and_complete_ne4_scream_cdf5.nc`

For P2x1 (ie with 2 MPI's), if I allow for core (`unlimit coredumpsize`), I see following:
```
#0  0x000014bb97bd18ef in _cray_mpi_memcpy_rome () from /opt/cray/pe/lib64/libmpi_gnu_91.so.12
#1  0x000014bb9602c54c in MPIR_Localcopy () from /opt/cray/pe/lib64/libmpi_gnu_91.so.12
#2  0x000014bb97a66b1b in MPIDI_CRAY_Common_lmt_unpack () from /opt/cray/pe/lib64/libmpi_gnu_91.so.12
#3  0x000014bb964fb9a3 in PMPI_Startall () from /opt/cray/pe/lib64/libmpi_gnu_91.so.12
#4  0x0000000002b236ea in Homme::BoundaryExchange::exchange (this=0x1f5e6dd0, rspheremp=0x0) at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/homme/src/share/cxx/mpi/BoundaryExchange.cpp:260
#5  0x0000000002b2359f in Homme::BoundaryExchange::exchange (this=0x1f5e6dd0) at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/homme/src/share/cxx/mpi/BoundaryExchange.cpp:232
#6  0x000000000278d5b7 in scream::PhysicsDynamicsRemapper::do_remap_fwd (this=0x1fda6600) at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/dynamics/homme/physics_dynamics_remapper.cpp:474
#7  0x000000000319cee7 in scream::AbstractRemapper::remap (this=0x1fda6600, forward=true) at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/share/grid/remap/abstract_remapper.cpp:128
#8  0x00000000026ec1fb in scream::HommeDynamics::initialize_homme_state (this=0x1bcc4690) at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:942
#9  0x00000000026e4190 in scream::HommeDynamics::initialize_impl (this=0x1bcc4690, run_type=scream::RunType::Initial)
    at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:349
#10 0x0000000003133dd5 in scream::AtmosphereProcess::initialize (this=0x1bcc4690, t0=..., run_type=scream::RunType::Initial)
    at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/share/atm_process/atmosphere_process.cpp:20
#11 0x000000000314403d in scream::AtmosphereProcessGroup::initialize_impl (this=0x18f119e0, run_type=scream::RunType::Initial)
    at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/share/atm_process/atmosphere_process_group.cpp:139
#12 0x0000000003133dd5 in scream::AtmosphereProcess::initialize (this=0x18f119e0, t0=..., run_type=scream::RunType::Initial)
    at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/share/atm_process/atmosphere_process.cpp:20
#13 0x0000000002546fb1 in scream::control::AtmosphereDriver::initialize_atm_procs (this=0x1ba0a820) at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/control/atmosphere_driver.cpp:775
#14 0x00000000005c0e01 in <lambda()>::operator()(void) const (__closure=0x7ffc8ab09180) at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp:194
#15 0x00000000005c5cd5 in _GLOBAL__N__59_tmpxft_00021865_00000000_7_scream_cxx_f90_interface_cpp1_ii_babe8b2b::fpe_guard_wrapper<scream_init_atm(int const&, int const&)::<lambda()> >(const <lambda()> &) (f=...)
    at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp:50
#16 0x00000000005c0e5b in scream_init_atm (start_ymd=@0x7ffc8ab0921c: 10101, start_tod=@0x7ffc8ab09218: 0) at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp:170
#17 0x00000000005bd858 in atm_comp_mct::atm_init_mct (eclock=..., cdata=..., x2a=..., a2x=..., nlfilename=..., _nlfilename=6)
    at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/atm_comp_mct.F90:152
#18 0x000000000044337f in component_mod::component_init_cc (eclock=..., comp=..., comp_init=0x5bcfbb <atm_comp_mct::atm_init_mct>, infodata=..., nlfilename=..., seq_flds_x2c_fluxes=..., seq_flds_c2x_fluxes=..., 
    _nlfilename=6, _seq_flds_x2c_fluxes=0, _seq_flds_c2x_fluxes=0) at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/component_mod.F90:248
#19 0x000000000042b1d3 in cime_comp_mod::cime_init () at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_comp_mod.F90:1425
#20 0x000000000043c5ac in cime_driver () at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_driver.F90:122
#21 0x000000000043c6ec in main (argc=1, argv=0x7ffc8ab0c16f) at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_driver.F90:23
#22 0x000014bb9440f34a in __libc_start_main () from /lib64/libc.so.6
#23 0x0000000000412fba in _start () at ../sysdeps/x86_64/start.S:120
```

**Comment by bartgol** [comment]
Actually, this wasn't that hard to track down. @tcclevenger I think [this](https://github.com/E3SM-Project/scream/blob/d0864c3b0fa5e35cf75e2eed3ffaba2b4a26fe98/components/scream/src/control/surface_coupling_export.cpp#L46) line is the problem (or at least part of it). That policy is used in `calculate_z_int`, which requires `column_scan`. The policy should be created with the special scan-safe ekat util.

**Comment by bartgol** [comment]
@ndkeen I pushed a mini-branch. You can cherry-pick the commit, and see if it fixes things for you.

**Comment by tcclevenger** [comment]
@bartgol Yep, this was my issue/fix as well. I'm creating a PR.

**Comment by bartgol** [comment]
Ok, I'm deleting my branch then. Noel can simply grab the fix(es) from your PR.

**Comment by ndkeen** [comment]
With that one-line change, I do see a diff error now: ```
0: A100-SXM4-40GB
0:  Reading setup_nml
0:  Reading grid_nml
0:  Reading ice_nml
0:  Reading tracer_nml
0: CalcWorkPerBlock: Total blocks:          1 Ice blocks:          1 IceFree blocks:          0 Land blocks:          0
0: Atmosphere step = 0; model time = 0001-01-01 00:00:00
0: terminate called after throwing an instance of 'std::logic_error'
0:   what():  /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:407: FAIL:
0: false
0: Kokkos::Impl::ParallelFor< Cuda > requested too large team size.
0: Traceback functionality not available
0:
0:
0: Program received signal SIGABRT: Process abort signal.
0:
0: Backtrace for this error:
0: #0  0x1535d998249f in ???
0: #1  0x1535d9982420 in ???
0: #2  0x1535d9983a00 in ???
0: #3  0x1535dfcebf59 in _ZN9__gnu_cxx27__verbose_terminate_handlerEv
0:      at ../../../../cpe-gcc-11.2.0-202108140355.9bf1fd589a5c1/libstdc++-v3/libsupc++/vterminate.cc:95
0: #4  0x1535dfcf77b9 in _ZN10__cxxabiv111__terminateEPFvvE
0:      at ../../../../cpe-gcc-11.2.0-202108140355.9bf1fd589a5c1/libstdc++-v3/libsupc++/eh_terminate.cc:48
0: #5  0x1535dfcf7824 in _ZSt9terminatev
0:      at ../../../../cpe-gcc-11.2.0-202108140355.9bf1fd589a5c1/libstdc++-v3/libsupc++/eh_terminate.cc:58
0: #6  0x1535dfcf7ab7 in __cxa_throw
0:      at ../../../../cpe-gcc-11.2.0-202108140355.9bf1fd589a5c1/libstdc++-v3/libsupc++/eh_throw.cc:95
0: #7  0x26e514d in _ZN6scream13HommeDynamics8run_implEi
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:407
0: #8  0x3134805 in _ZN6scream17AtmosphereProcess3runEi
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/share/atm_process/atmosphere_process.cpp:46
0: #9  0x3144419 in _ZN6scream22AtmosphereProcessGroup14run_sequentialEd
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/share/atm_process/atmosphere_process_group.cpp:158
0: #10  0x3144333 in _ZN6scream22AtmosphereProcessGroup8run_implEi
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/share/atm_process/atmosphere_process_group.cpp:145
0: #11  0x3134805 in _ZN6scream17AtmosphereProcess3runEi
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/share/atm_process/atmosphere_process.cpp:46
0: #12  0x25475b0 in _ZN6scream7control16AtmosphereDriver3runEi
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/control/atmosphere_driver.cpp:817
0: #13  0x5c0e8f in operator()
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp:204
0: #14  0x5c5d10 in fpe_guard_wrapper<scream_run(const Real&)::<lambda()> >
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp:50
0: #15  0x5c0eb2 in scream_run
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp:201
0: #16  0x5bcde0 in __atm_comp_mct_MOD_atm_run_mct
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/mct_coupling/atm_comp_mct.F90:209
0: #17  0x43efcb in __component_mod_MOD_component_run
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/component_mod.F90:728
0: #18  0x424125 in __cime_comp_mod_MOD_cime_run
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_comp_mod.F90:3082
0: #19  0x43c684 in cime_driver```

**Comment by tcclevenger** [comment]
Hmm I'm not seeing that on my CIME summit runs. I'm only running with `P1x1`. I wonder if that could trigger this.

**Comment by bartgol** [comment]
For the Nx1 case, I suspect that you see the core dump of the ranks that did not run into the kernel-too-big exception. Those ranks seem to receive the abort command while they're stuck waiting in a collective for the rank that crashed. I think sticking with the 1x1 case is the best way to debug this.

The last dump you posted is likely just from the program terminating due to uncaught exception, and likely hitting another issue. Destructors in C++ should be implemented exception-free, but that's often not the case. C++ does not support a second exception being thrown while the stack is unwind due to a previous exception, and, if that happens the application just terminates. The dump you get in that case might not reflect what you expected, or what the real issue is.

I would run 1x1 with DEBUG=TRUE, through gdb, and set both `catch throw` as well as `b MPI_Abort`, so that as soon as a fatal error happens, execution stops (in case of exception, before, and this is important, stack unwinding starts), and we can clearly see what the original error is.

**Comment by ndkeen** [comment]
Does this problem run on other machines?

**Comment by bartgol** [comment]
@tcclevenger what case/compset are you running on Summit? Is it F2010-SCREAMv1-AQP1?

**Comment by ndkeen** [comment]
Wouldn't the issue be `0:  ERROR: ice: Vertical thermo error` ?

**Comment by tcclevenger** [comment]
My test case is 
```
SMS_D_Ln2_P1x1.ne4_ne4.F2000-SCREAMv1-AQP1.summit_gnugpu.scream-30min_ts
```

**Comment by bartgol** [comment]
> Wouldn't the issue be `0: ERROR: ice: Vertical thermo error` ?

Ah, yeah, that could be the problem. Is there a v0 equivalent of the v1 testcases? @AaronDonahue , do you know? If not, then we should work on making our v1 testcases an exact mirror of the v0 ones, to avoid errors due to other factors.

**Comment by PeterCaldwell** [comment]
I'm a bit confused here. Haven't we been running SMS_D_Ln2_P96x1.ne4_ne4.F2000SCREAMv1.quartz_intel on CPU machines for at least a month (e.g. https://acme-climate.atlassian.net/wiki/spaces/NGDNA/pages/3243900990/SCREAMv1+runs+on+LC+quartz)? Why would we need a v0 equivalent when we already have this config working in v1 on other machines?

**Comment by bartgol** [comment]
@PeterCaldwell I don't advocate for creating v0 equivalents to our v1 cases. I am more for creating our v1 cases to match existing v0 compsets.

We are still developing/fixing v1, and some mods will generated non bfb changes. If one change ends up making the model crash, I want to be sure it is our fault, and not, for instance, a bad set of input params/fields. If the compset matches a working one in v0, we can be absolutely sure that any crash is 100% our fault.

In noel case, the error message seems to suggest that some field in CICE goes bananas. That _might_ be our fault...or it might be a compset issue.

If the same version of SCREAM that noel is using is running on quartz, it is likely that we are doing something wrong on our end (perhaps something GPU-related, which causes some scream exported flux to be wrong). But the problem stands: if in a week F2000SCREAMv1 starts failing, can we be 100% sure it is our fault, and not maybe a problem with the parameters which for some lucky reason ended up not causing crashes till now?

**Comment by AaronDonahue** [comment]
@PeterCaldwell - we have nightly tests on mappy and quartz for both of the FSCREAMv1 compsets (normal and aquaplanet).  Both of these are CPU machines.  We do not test on a GPU machine and as far as I know we haven't run on a GPU machine either.  @tcclevenger, @brhillman or @jgfouca can correct me if I'm wrong.

I agree with @bartgol though that as long as our parameter space itself isn't tested we could end up chasing bugs that are not really bugs.  We learned this when I was able to get the run passing just by changing the HOMME params.  We should probably try to match the run setup of v0 as much as possible.  I.e. timestep, homme params, physics params.  It won't be a perfect match but should at least rule out some potential causes.

**Comment by bartgol** [comment]
Ah, good point about the outstanding HOMME params issue. And in general, we do not have the same way to set defaults as v0 (as per our email conversation).

It might be complicated to have _the same_ compset running in v0 and v1, but we should try to do what we can to make them as close as possible, while we work out our parameters and defaults settings issues.

**Comment by AaronDonahue** [comment]
Next is still closed on the E3SM repo, so it will at least a couple more days before I can merge the homme namelist PR.  Once that goes in though we should be able to match the homme settings from SCREAMv0 much more easily.

**Comment by tcclevenger** [comment]
@ndkeen I have a fix for the `P1x1` error (not calling `Kokkos::Cuda::finalize()`). It is the only commit on my branch `tcclevenger/cime_on_summit_fixes`

**Comment by PeterCaldwell** [comment]
Ok, thanks for the discussion. Since Aaron says this v1 SPS_D test is passing nightly on quartz and mappy, I think in this specific instance the problem must be GPU or Perlmutter-specific rather than a bad input file (unless Noel did something non-standard to the case). Correct me if I'm wrong.

But I totally agree that we should have v0 and v1 tests which are as similar as possible. My strategy here was to not worry about making v0 and v1 identical until we fixed the obvious pathologies in v1. Kind of like how you don't need an electron microscope to distinguish between a diamond and a piece of coal. Let's use what we have until we get v0 and v1 close enough together that inputdata differences are the leading source of discrepancies.

**Comment by ndkeen** [comment]
This is only change?  `components/scream/src/mct_coupling/scream_cxx_f90_interface.cpp`
```
>     // Finalize scream session
>     scream::finalize_scream_session();
```

**Comment by tcclevenger** [comment]
Yes. The issue is `Kokkos::Cuda::finalize()` was never called, and this call eventually get there. For me I got no other errors for `P1x1`.

**Comment by tcclevenger** [comment]
For the `PNx1` case, I am getting this error
```
1: 1: errno 14 pid 3311010
1: 1: e3sm.exe: /__SMPI_build_dir__________________________/ibmsrc/pami/ibm-pami/buildtools/pami_build_port/../pami/components/devices/shmem/shaddr/CMAShaddr.h:164: size_t PAMI::Device::Shmem::CMAShaddr::read_impl(PAMI::Memregion*, size_t, PAMI::Memregion*, size_t, size_t, bool*): Assertion `cbytes > 0' failed.
1: 1: 
1: 1: Program received signal SIGABRT: Process abort signal.
1: 1: 
1: 1: Backtrace for this error:
1: 0: errno 14 pid 3311011
1: 0: e3sm.exe: /__SMPI_build_dir__________________________/ibmsrc/pami/ibm-pami/buildtools/pami_build_port/../pami/components/devices/shmem/shaddr/CMAShaddr.h:164: size_t PAMI::Device::Shmem::CMAShaddr::read_impl(PAMI::Memregion*, size_t, PAMI::Memregion*, size_t, size_t, bool*): Assertion `cbytes > 0' failed.
1: 0: 
1: 0: Program received signal SIGABRT: Process abort signal.
1: 0: 
1: 0: Backtrace for this error:
ERROR:  One or more process (first noticed rank 1) terminated with signal 6 (core dumped)
```
Doesn't seem helpful at this point. But looks like it occurs in `scream::HommeDynamics::initialize_impl` which seems to be where your run errors out.

**Comment by tcclevenger** [comment]
@PeterCaldwell It is just the 2 steps like in the mappy SCREAMv1 run.

**Comment by tcclevenger** [comment]
@PeterCaldwell Should I run it out as far as I can?

**Comment by PeterCaldwell** [comment]
@tcclevenger - ok cool. I guess you might as well ask it to run for a month and see how far it gets before timing out or crashing just because it would be easy to do. No hurry on that though - figuring out crashes on multiple threads is more important.

**Comment by ndkeen** [comment]
With the above finalize change, the error I see is:

```
0:  calling getglobalwrite with decomp_index=          503  and elmlevel= landunit
0:  ERROR: Unknown error submitted to shr_abort_abort.
0: #0  0x21640a5 in __shr_abort_mod_MOD_shr_abort_backtrace
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/share/util/shr_abort_mod.F90:104
0: #1  0x216425b in __shr_abort_mod_MOD_shr_abort_abort
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/share/util/shr_abort_mod.F90:61
0: #2  0x6012b7 in __abortutils_MOD_endrun_globalindex
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/elm/src/main/abortutils.F90:77
0: #3  0x10ca697 in net_longwave
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/elm/src/biogeophys/UrbanRadiationMod.F90:470
0: #4  0x10dd0bc in __urbanradiationmod_MOD_urbanradiation
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/elm/src/biogeophys/UrbanRadiationMod.F90:245
0: #5  0x61ed59 in __elm_driver_MOD_elm_drv
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/elm/src/main/elm_driver.F90:1316
0: #6  0x5efea0 in __lnd_comp_mct_MOD_lnd_run_mct
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/components/elm/src/cpl/lnd_comp_mct.F90:512
0: #7  0x43f1bc in __component_mod_MOD_component_run
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/component_mod.F90:728
0: #8  0x42339a in __cime_comp_mod_MOD_cime_run
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_comp_mod.F90:2881
0: #9  0x43c874 in cime_driver
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_driver.F90:153
0: #10  0x43c8d7 in main
0:      at /pscratch/sd/n/ndk/wacmy/s12-mar1/driver-mct/main/cime_driver.F90:23

in lnd.log:
ENDRUN:ERROR in /pscratch/sd/n/ndk/wacmy/s12-mar1/components/elm/src/biogeophys/UrbanRadiationMod.F90 at line 470

which is in subroutine net_longwave()

         err = lwdown(l) - (lwdown_road(l) + (lwdown_shadewall(l) + lwdown_sunwall(l))*canyon_hwr(l))
         if (abs(err) > 0.10_r8 ) then
#ifndef _OPENACC
             write(iulog,*) 'urban incident atmospheric longwave radiation balance error',err
             write(iulog,*) 'l          = ',l
             write(iulog,*) 'lwdown     = ',lwdown(l)
             write(iulog,*) 'vf_sr      = ',vf_sr(l)
             write(iulog,*) 'vf_sw      = ',vf_sw(l)
             write(iulog,*) 'canyon_hwr = ',canyon_hwr(l)
             write(iulog,*) 'elm model is stopping'
             call endrun(decomp_index=l, elmlevel=namel, msg=errmsg(__FILE__, __LINE__)) ! <-- line 470
#endif
         endif
      end do

```

**Comment by bartgol** [comment]
Random thought I just got: when we do the import/export, our scream fields are partitioned by MPI rank. Are the arrays passed by the coupler rank-specific or are they global? Meaning, are we supposed to read/write at a particular, rank-specific, location in the array?

**Comment by tcclevenger** [comment]
We are definitely treating it in `surface_coupling` as if they are rank specific.

**Comment by ndkeen** [comment]
What is the mechanism for me to run the same test without gpu? That is, `SMS_D_Ln2_P1x1.ne4_ne4.F2000SCREAMv1.perlmutter_gnu` ? As currently, the machine settings seem to only allow one or the other. Would we have two machine names? `perlmutter-gpu` and `perlmutter-cpu` ?

**Comment by bartgol** [comment]
Ah, yeah, currently SCREAM uses the cime machine name to load a cmake machine file for kokkos. In a pinch, you can replace the perlmutter.cmake mach file with something that is good for the particular CPU you want.

But long term, yes, I think we will have to create two mach files, or add some cmake cache var to tell scream which arch to build for.

**Comment by ndkeen** [comment]
Pinching, I see that `SMS_D_Ln2_P1x1.ne4_ne4.F2000SCREAMv1.perlmutter_gnu` does complete the 2 ATM steps.

**Comment by bartgol** [comment]
does perlmutter_gpu use the same host compiler used by perlmutter_gnu?

**Comment by ndkeen** [comment]
I'm not sure your question. For e3sm, we have a machine name and then different compilers, such as `gnu` and `gnugpu` (though I personally think we should leave the compiler `gnu` and create another variable to indicate the HW to target, but that's diff conversation). For both of those, we are free to use whatever modules and flags we chose -- but in this case, it's the same.

**Comment by ndkeen** [comment]
I noticed that there isn't much written to atm.log file. Is that intended?  Even with the job that did complete, the only line written is: `P3_INIT (reading/creating look-up tables) ...`

**Comment by bartgol** [comment]
I mean, when you use perlmutter_gpu, nvcc is used to compile CUDA, but something else is used to compile normal host code. I am guessing that's again gnu, but I am not sure if it's the same version.

**Comment by PeterCaldwell** [comment]
"I noticed that there isn't much written to atm.log file" - Yeah, all our output is being written to e3sm.log instead for some reason.

**Comment by bartgol** [comment]
> "I noticed that there isn't much written to atm.log file" - Yeah, all our output is being written to e3sm.log instead for some reason.

That's probably my fault. I might have messed up something when I implemented our atm_comp_mct.F90 module... I will double check.

**Comment by ndkeen** [comment]
Same GNU compiler version for host/device.  I can see the ATM step prints in e3sm.log, so that's fine.

I'm back to getting `Kokkos::Impl::ParallelFor< Cuda > requested too large team size.` for DEBUG.
Adding print, I see that for the line: `if (int(m_team_size) > blah`, it is 16 > 12.

**Comment by bartgol** [comment]
That's an annoying error. We need to verify which Functor is too big, and add launch bounds for that functor.

**Comment by ndkeen** [comment]
Above I noted that I hit `ERROR: ice: Vertical thermo error` in DEBUG.  However, in other DEBUG cases, I hit the `Kokkos::Impl::ParallelFor< Cuda > requested too large team size.` error. I now think it's "sometimes" stopping with thermo error before going on. In same case, it stopped the thermo error and simply resubmitting stopped with team size error.  Regarding the team size, I added a silly printf to that location to see what all of the sizes are and I got this:

```
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=   96 opt_team_size=   64 max_team_size= 1024
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=   96 opt_team_size=   96 max_team_size= 1024
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    4 max_team_size=   64
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=  256 opt_team_size=   32 max_team_size=  384
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=   32 opt_team_size=   96 max_team_size= 1024
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=   64 opt_team_size=   32 max_team_size=  512
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=   64 opt_team_size=   32 max_team_size=  768
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=   32 opt_team_size=   96 max_team_size= 1024
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=   32 opt_team_size=   64 max_team_size= 1024
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=   64 opt_team_size=   96 max_team_size= 1024
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=   64 opt_team_size=   32 max_team_size=  768
0: ndk m_vector_size=  1 arg_policy.impl_vector_length()=  1 team_size=  256 opt_team_size=   32 max_team_size=  384
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   40
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    4 max_team_size=   64
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   40
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    4 max_team_size=   64
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    4 max_team_size=   64
0: ndk m_vector_size=  2 arg_policy.impl_vector_length()=  2 team_size=   72 opt_team_size=   16 max_team_size=  448
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   48
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    4 max_team_size=   64
0: ndk m_vector_size=  2 arg_policy.impl_vector_length()=  2 team_size=   72 opt_team_size=   16 max_team_size=  448
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   48
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    4 max_team_size=   64
0: ndk m_vector_size=  2 arg_policy.impl_vector_length()=  2 team_size=   72 opt_team_size=   16 max_team_size=  448
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   48
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=   10 max_team_size=   64
0: ndk m_vector_size= 16 arg_policy.impl_vector_length()= 16 team_size=   16 opt_team_size=    2 max_team_size=   24
0: ndk m_vector_size= 32 arg_policy.impl_vector_length()= 32 team_size=   16 opt_team_size=    1 max_team_size=   32
0: ndk m_vector_size= 32 arg_policy.impl_vector_length()= 32 team_size=   16 opt_team_size=    2 max_team_size=   32
0: ndk m_vector_size= 32 arg_policy.impl_vector_length()= 32 team_size=   16 opt_team_size=    1 max_team_size=   16
0:   what():  /pscratch/sd/n/ndk/wacmy/s12-mar1/components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:407: FAIL:
0: ndk m_vector_size= 32 arg_policy.impl_vector_length()= 32 team_size=   16 opt_team_size=    1 max_team_size=   12
0: ndk ekat/extern/kokkos/core int(m_team_size)=16 blah=12
```

**Comment by ndkeen** [comment]
Also meant to report that I can a full day without DEBUG!

**Comment by ndkeen** [comment]
To try with valgrind on PM, I did:

```
module load valgrind4hpc
login26% which valgrind
/opt/cray/pe/valgrind4hpc/2.12.7/bin/valgrind
login26% valgrind --version
valgrind-3.16.0.RC2

edit env_mach_specific.xml and modify:
<entry id="run_exe" value="valgrind  --error-limit=no --log-file=log --leak-check=yes ${EXEROOT}/e3sm.exe ">

./case.submit

```

Running a DEBUG case again with that, it's obv running much slower and is not complete (will have to resubmit for longer) , but here is the valgrind output so far: `/global/cfs/cdirs/e3sm/ndk/vlog-incomplete.txt`

**Comment by ndkeen** [comment]
I have already been running with gnu v10, but I also updated other modules and adjusted fortran flags on scream side. Running again with `P1x1.ne4_ne4.F2000SCREAMv1.perlmutter_gnugpu` I see that with DEBUG, can complete 2 steps (as before). 

And without DEBUG, still hit:
`0: Negative (or nan) layer thickness detected, aborting!`

For both of those cases I resubmitted in same case and see the same results.

**Comment by ndkeen** [comment]
Using GNU v11, I see same behavior -- runs 2 steps in DEBUG and fails same way with optimized build.

**Comment by tcclevenger** [comment]
On weaver, we can run aquaplanet with multiple ranks error free (I tested up to 300 steps). Here is the testcase: `SMS_D_Ln2_P4x1.ne4_ne4.F2000-SCREAMv1-AQP1.weaver_gnugpu.scream-30min_ts`. The way the job was submitted seemed bugged (I think the `mpirun` command was not correct), so I ran manually using the following after building:
```
cd $case_dir/run/
mkdir -p timing/checkpoints/ && touch timing/checkpoints/model_timing_00010101_00000_stats

mpirun --map-by ppr:2:socket --bind-to core $case_dir/bld/e3sm.exe < drv_in
```

Since weaver and summit are both V100 GPUs, I think the code must be alright (at least for aquaplanet). I'll keep investigating summit. My guess is that something is off with the `jsrun` vars used, but that is not anything but a guess.

**Comment by ndkeen** [comment]
Good news that you're able to run with more than 1 MPI.

Regarding my issues on PM, after some ugly debugging, I have learned something that is perhaps not all that useful. If I build optimized, but then set DEBUG for kokkos only, the 2-step test case I'm trying will complete. And if I build DEBUG, but turn on kokkos debugging, it also stops in the same way.

I had been working with a repo from March 1st, but then to confirm this with updated source, I tried repo of March 12th and ran into a new issue (https://github.com/E3SM-Project/scream/issues/1474).  But with a repo just before the last rad-based PR, I am able to repeat my findings.

**Comment by tcclevenger** [comment]
Found one way to fix the MPI issue with summit:add `--smpiargs "-gpu"` to the `jsrun` command. @ndkeen does PM use `jsrun`?

**Comment by ndkeen** [comment]
No PM doesn't use that. But because adjusting this sort of thing was having an impact, I revisited some of the slurm gpu settings that I had set up during my HOMME stand-alone testing (and also work with MMF). Sure enough, I'm able to get MPI runs with gpus by making adjustments that I don't think should matter. More later.

**Comment by PeterCaldwell** [comment]
So it sounds like we can run ne4 on both Summit and Perlmutter GPUs using multiple MPIs and multiple threads now? Does this mean the "run on GPU" task is finished (minus cleanup and sorting out last details like figuring out why the needed flags are needed, which may take a while)? In particular, I'm wondering whether we have a config for Hannah (whose github name I can't find) to start performance profiling?

**Comment by tcclevenger** [comment]
Yes, I am about to try non-aqua. But I think this is an issue on CPU as well? Can't remember..

**Comment by ndkeen** [comment]
Yes, I'm assuming this issue is about `F2000SCREAMv1` specifically, but I'm also testing things with aquaplanet.  I did try aquaplanet on CPU-only and it ran for 5 days (using 32x2).

**Comment by bartgol** [comment]
Isn't one of our v1 tests in the nightlies using topography?

**Comment by tcclevenger** [comment]
I think it will fail if more than the 2 time steps are done. I'm testing that now.

**Comment by ndkeen** [comment]
Yea I was thinking of a table, but things are moving fast enough (stuff getting fixed!) that I wanted to let dust settle.
One start:

```
           ne4 F2000-SCREAMv1-AQP1  ne4 F2000SCREAMv1
DEBUG               OK                    OK
OPT w KD            OK                    OK
OPT               neg layer            neg layer

where KD= changes to enable only Kokkos DEBUG, everything else optimized

```

And "OK" means it can (or will) run 5 days with 4 MPI's and GPU's.

**Comment by tcclevenger** [comment]
On Mappy, `ne4 F2000SCREAMv1` seems to be fine. Job is still in the queue on summit.

**Comment by AaronDonahue** [comment]
> Isn't one of our v1 tests in the nightlies using topography?

@bartgol , yes that is correct.  And Peter and I have run on quartz with topo for at least a month of simulation time.  Granted that was at ne4 resolution.

**Comment by AaronDonahue** [comment]
Here's a link to the test in our nightlies: https://my.cdash.org/testSummary.php?project=828&name=SMS_D_Ln2_P24x1.ne4_ne4.F2000SCREAMv1.mappy_gnu9.scream-30min_ts&date=2022-03-16

**Comment by tcclevenger** [comment]
Looks like we can up the number of steps if we want. Mappy is up to 90 iters for me now in my test.

**Comment by PeterCaldwell** [comment]
Ok, so it sounds like the negative layer thickness errors Noel is seeing are unique to Perlmutter (assuming Summit can run) and aren't related to topography. 

Noel - can you run the same case that gives negative layer thickness but just using Perlmutter CPUs? If so, then the problem is related to GPU usage(?).

**Comment by bartgol** [comment]
> Yea I was thinking of a table, but things are moving fast enough (stuff getting fixed!) that I wanted to let dust settle.

That's a good idea. As soon as we work out some kinks to get more consistent results, we should start a confluence page, with all the details regarding how to run on PM and SM, and what works and what doesn't.

**Comment by ndkeen** [comment]
Another attempt at summary after trying a few more things:
```
                    ne4 F2000-SCREAMv1-AQP1            ne4 F2000SCREAMv1
GPU DEBUG               OK                          Tl1_1 has <= 0 at 5 steps
GPU OPT w KD            OK                          Tl1_1 has <= 0 at 5 steps
GPU OPT             neg layer before step 2         neg layer before step 2

CPU DEBUG               OK                            error at 5 steps
CPU OPT w KD            OK                            error at 5 steps
CPU OPT                 OK                            error at 5 steps

where KD= changes to enable only Kokkos DEBUG, everything else optimized
```

The "error at 5 steps" for the CPU attempts may be same issue as we see with GPU, but it doesn't hit the same assert.

**Comment by bartgol** [comment]
We should build our homme+physics test (i.e., not a CIME run), and run it with the same input file used by ne4 F2000SCREAMv1, to see if we can reproduce the Tl1_1/thickness errors "in house". It will also tell us if the issue is potentially our case config (e.g., "bad" nml options, when coupling to other components) or more scream-related bugs.

**Comment by tcclevenger** [comment]
`F2000SCREAMv1-AQP1`  is fine on Summit (I tested up to 50 step, but it goes much farther) for both DEBUG and OPT. For `F2000SCREAMv1` I get both Debug and OPT fails. Somewhere outside of SCREAM we get 
```
1: 1: Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.
```
Is there something to do with topography that other models are expecting to be exported in the SurfaceCoupling?

**Comment by tcclevenger** [comment]
Never mind, I also only get the error in OPT.

**Comment by ndkeen** [comment]
I was ignorantly running with default timestep (NCPL=12) and this was reason for the errors after step 5 for non-aquaplanet.  When I run again with NCPL=48 (though Peter informs me I should actually use NCPL=288), these tests now complete 5 days.

```
                    ne4 F2000-SCREAMv1-AQP1          ne4 F2000SCREAMv1
GPU DEBUG               OK                                OK
GPU OPT w KD            OK                                OK
GPU OPT             need to try again               need to try again

CPU DEBUG               OK                                OK
CPU OPT w KD            OK                                OK
CPU OPT                 OK                          need to try again

where KD= changes to enable only Kokkos DEBUG, everything else optimized

**Comment by PeterCaldwell** [comment]
So I think this means that we can run ne4 on Perlmutter GPUs for both aquaplanet and realistic topography without any trouble, right? And I think you still weren't sure *why* the particular non-default settings you were using allowed us to do these runs? Did you ever sort that last question out?

**Comment by ndkeen** [comment]
Peter: well I still seem to have issues with true OPT builds, though the error is different than above (in fact, what might be the same as Conrad)

If I understand your 2nd question... I guess it doesn't matter why the default NCPL=12 *did* work, and maybe not interesting if it did or not -- we just want to use correct settings now.  Indeed the testnames that machines like quartz was using *does* use the modifier that changes NCPL=48.

As this issue has gotten long, and has many problems/fixes, I'm going to make another issue that is ne4 with NCPL=288

**Comment by PeterCaldwell** [comment]
Ok, sounds good. I didn't read your last report closely enough and didn't realize that you're still getting OPT fails.

**Comment by tcclevenger** [comment]
These errors make me think there is an issue in `SurfaceCoupling::do_export()`. I'm looking into it.

**Comment by tcclevenger** [comment]
I now trigger the error in OPT and DEBUG after updating my branch. The issue is in `scream/components/elm/src/biogeophys/SnowSnicarMod.F90:2372` we have 
```
tau(i)    = tau_sum + tau_snw(i)
omega(i)  = (1/tau(i))*(...blah...)
```
and for one of the iterations on `i`, `tau_sum = tau_snw(i) = 0`.

This is after the AD has run one iteration. I know nothing about ELM or what these tau vars are, but my guess is this is from a scream export variable being wrong, which could be an issue in SurfaceCoupling, but maybe more likely a variable being computed inside SCREAM with incorrect topology?

**Comment by PeterCaldwell** [comment]
This part of the code is dealing with radiative transfer so double check the radiation variables (SW, LW) that we pass to the surface. It also makes a lot of mention of aerosols, which we don't use at all. If the problem is that we've zeroed out all of the aerosol info we pass to the surface, I'm surprised v0 would run though...

**Comment by bartgol** [comment]
The case pack_size=1, with a scan from the top is actually the simplest one, when you look at the impl. I went over it, but couldn't spot any suspicious thing.

**Comment by ndkeen** [comment]
I'm going to close this general catch-all issue as we are much further along running cime cases at ne4/ne30 now.

**Comment by tcclevenger** [comment]
@ndkeen With the fixes in https://github.com/E3SM-Project/scream/pull/1526 (which should automerge soon) I was able to run on Summit for 1/2 a month (It only timed out, so should go farther). Not sure if this will fix the issue you are seeing, but it might.
